File: CrackingTheCodingInterview/ch04-TreesAndGraphs/checkSubtree.js
/*
  T1 and T2 are two very large binary trees, with T1 much bigger
  than T2 - create an algorithm to determine if T2 is a subtree
  of T1
*/

// traverse through T1
// each time node is found that matches T2 root,
// call matchTree to compare the two subtrees
function checkSubtree(T1, T2) {
  if (!T2) return true;
  if (!T1) return false;
  if (T1.value === T2.value && matchTree(T1, T2)) {
    return true;
  }

  return checkSubtree(T1.left, T2) || checkSubtree(T1.right, T2);
}

function matchTree(a, b) {
  if (!a && !b) { // nothing left in the subtree
    return true;
  } else if (!a || !b) { // exactly one tree is empty, therefore no match
    return false;
  } else if (a.value !== b.value) { // values don't match
    return false;
  } else {  // check both sides
    return matchTree(a.left, b.left) && matchTree(a.right, b.right);
  }
}

File: CrackingTheCodingInterview/ch04-TreesAndGraphs/checkSubtree.py
"""
T1 and T2 are two very large binary trees, with T1 much bigger
than T2 - create an algorithm to determine if T2 is a subtree
of T1
"""


# Traverse t1 until find value equal to t2, then check match_trees
def is_subtree(t1, t2):
    if not t2:
        return True

    if not t1:
        return False

    if match_trees(t1, t2):
        return True

    return is_subtree(t1.left, t2) or \
        is_subtree(t1.right, t2)


# Check if trees match exactly
def match_trees(t1, t2):
    # Both subtrees empty
    if not t1 and not t2:
        return True

    # One subtree empty
    if not t1 or not t2:
        return False

    # Values don't match
    if t1.value !== t2.value:
        return False

    # Check both sides
    return match_trees(t1.left, t2.left) and \
        match_trees(t1.right, t2.right)

